src/Microsoft.AspNetCore.OData/Query/IQueryableODataExtension.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                return (TResult)Convert.ChangeType(source, typeof(TResult), CultureInfo.InvariantCulture);
+            }
+
+            return default;
+        }
+
+        /// <summary>
+        /// Converts the elements of an <see cref="IQueryable"/> to the specified type.
+        /// </summary>
+        /// <typeparam name="TResult">The type to convert the elements of source to.</typeparam>
+        /// <param name="source">The <see cref="IQueryable"/> that contains the elements to be converted.</param>
+        /// <param name="options">The cast options.</param>
+        /// <returns>Contains each element of the source sequence converted to the specified type.</returns>
+        [SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "<Pending>")]
+        public static IEnumerable<TResult> OCast<TResult>(this IQueryable source, ODataCastOptions options = null)
+        {
+            if (source is null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            foreach (var item in source)
+            {
+                yield return OCast<TResult>(item, options);
+            }
+        }
+
+        /// <summary>
+        /// Create the object based on <see cref="SelectExpandWrapper"/>
+        /// </summary>
+        /// <param name="wrapper">The select expand wrapper.</param>
+        /// <param name="resultType">The created type.</param>
+        /// <param name="options">The options.</param>
+        /// <returns>The created object.</returns>
+        private static object CreateInstance(SelectExpandWrapper wrapper, Type resultType, ODataCastOptions options)
+        {
+            object instance;
+            Type instanceType = resultType;
+            IEdmModel model = wrapper.Model;
+            if (wrapper.UseInstanceForProperties)
+            {
+                instance = wrapper.InstanceValue;
+                instanceType = instance.GetType();
+            }
+            else
+            {
+                if (wrapper.InstanceType != null && wrapper.InstanceType != resultType.FullName)
+                {
+                    // inheritance
+                    IEdmTypeReference typeReference = wrapper.GetEdmType();
+                    instanceType = model.GetClrType(typeReference); // inheritance type
+                }
+
+                instance = Activator.CreateInstance(instanceType);
+            }
+
+            IDictionary<string, object> properties;
+            if (options != null && options.MapProvider != null)
+            {
+                properties = wrapper.ToDictionary(options.MapProvider);
+            }
+            else
+            {
+                properties = wrapper.ToDictionary();
+            }
+
+            foreach (var property in properties)
+            {
+                string propertyName = property.Key;
+                object propertyValue = property.Value;
```

Comment:
Since `property` is a `KeyValuePair`:
```suggestion
            foreach (var (propertyName, propertyValue) in properties)
            {
```